## Project Title
Quantra Retail Performance and Business Intelligence Report

## Background
Quantra Retail is a fictional UK-based retail company operating across 10 major cities including London, Manchester, Birmingham, Leeds, Bristol, Edinburgh, Glasgow, Liverpool, Sheffield, and Cardiff. The company offers products across 6 key categories. Electronics, Clothing, Groceries, Home & Garden, Beauty, and Sports through both online and in-store channels. 

As the business scales across regions, understanding sales performance, customer behaviour, and product profitability becomes critical for driving revenue growth and operational efficiency.


## Business Problem
Quantra Retail operates across multiple cities and product categories 
but lacks a unified view of its commercial performance. The leadership 
team cannot clearly identify which regions are underperforming, which 
product categories are most profitable, or how customer purchasing 
behaviour varies across demographics. Without these insights, resource 
allocation, promotional strategies, and inventory decisions are being 
made without data backing — creating a significant risk of lost revenue 
and missed growth opportunities.

## Project Aim
To transform Quantra Retail's raw transactional data into a clean, 
structured dataset that powers meaningful SQL analysis and an 
interactive Power BI dashboard — enabling the business to make 
faster, smarter, and more confident commercial decisions.

1. Clean and validate all datasets to ensure accuracy and consistency
2. Identify top performing product categories and regions by revenue
3. Understand customer demographics and purchasing behaviour
4. Uncover seasonal and monthly sales trends across 2024–2026
5. Compare online vs in-store channel performance
6. Segment customers by value to support targeted marketing strategies
7. Deliver actionable recommendations backed by data evidence

#### Dataset
- `products.csv` → 30 products across 6 categories (product_id, product_name, product category, unit_price)
- `customers.csv` → 1,000 UK customers (customer_id, customer name, city, age, gender, email, join date segment)
- `transactions.csv` → 10,000 retail transactions (transaction_id, customer_id, product_id, date, quantity, unit price, store type, payment method, total amount)

## Import libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Load dataset

In [2]:
products_df = pd.read_csv('../data/products.csv')
customers_df = pd.read_csv('../data/customers.csv')
transactions_df = pd.read_csv('../data/transactions.csv')

print("✅ All datasets loaded successfully!")
print(f"Products shape: {products_df.shape}")
print(f"Customers shape: {customers_df.shape}")
print(f"Transactions shape: {transactions_df.shape}")

✅ All datasets loaded successfully!
Products shape: (30, 5)
Customers shape: (1000, 7)
Transactions shape: (10000, 9)


## Exploratory Data Analysis

#### Explore product table 

In [3]:
print("=" * 40)
print("PRODUCTS DATASET")
print("=" * 40)
print(products_df.info())
print("\nMissing values:")
print(products_df.isnull().sum())
print("\nDuplicates:", products_df.duplicated().sum())
print("\nSample data:")
products_df.head()

PRODUCTS DATASET
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   product_id        30 non-null     object 
 1   product_name      30 non-null     object 
 2   product_category  30 non-null     object 
 3   unit_price        30 non-null     float64
 4   cost_price        30 non-null     float64
dtypes: float64(2), object(3)
memory usage: 1.3+ KB
None

Missing values:
product_id          0
product_name        0
product_category    0
unit_price          0
cost_price          0
dtype: int64

Duplicates: 0

Sample data:


,product_id,product_name,product_category,unit_price,cost_price
0,P001,Laptop,Electronics,321.52,131.02
1,P002,Smartphone,Electronics,141.14,65.91
2,P003,Tablet,Electronics,369.55,222.84
3,P004,Headphones,Electronics,446.63,190.30
4,P005,Smart Watch,Electronics,213.85,87.45


#### Product Insight
- Dataset contains 30 products across 6 categories — clean and complete
- No missing values or duplicates detected
- unit_price ranges from **£5** to **£500** covering budget to premium products


#### Explore customers table 

In [4]:
print("=" * 40)
print("CUSTOMERS DATASET")
print("=" * 40)
print(customers_df.info())
print("\nMissing values:")
print(customers_df.isnull().sum())
print("\nDuplicates:", customers_df.duplicated().sum())
print("\nAge range:")
print(f"Min age: {customers_df['age'].min()}")
print(f"Max age: {customers_df['age'].max()}")
print("\nSample data:")
customers_df.head()

CUSTOMERS DATASET
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    1000 non-null   object
 1   customer_name  1000 non-null   object
 2   city           1000 non-null   object
 3   age            1000 non-null   int64 
 4   gender         1000 non-null   object
 5   email          1000 non-null   object
 6   join_date      1000 non-null   object
dtypes: int64(1), object(6)
memory usage: 54.8+ KB
None

Missing values:
customer_id      0
customer_name    0
city             0
age              0
gender           0
email            0
join_date        0
dtype: int64

Duplicates: 0

Age range:
Min age: 18
Max age: 70

Sample data:


,customer_id,customer_name,city,age,gender,email,join_date
0,C0001,Miss Judith Ali,Sheffield,32,Female,graeme66@example.com,2023-11-01
1,C0002,Lawrence Roberts,London,32,Male,dorothybartlett@example.com,2023-09-07
2,C0003,Eileen Taylor,Edinburgh,43,Female,pricesharon@example.org,2023-07-08
3,C0004,Katherine Taylor,Manchester,31,Female,kathleenbevan@example.com,2026-06-03
4,C0005,Marilyn Clark,Leeds,59,Female,hewittmandy@example.com,2024-02-04


##### Customer Insight
- 1,000 UK customers across 10 major cities
- No missing values or duplicates detected
- Age range of **18–70** represents a broad demographic spread
- join_date stored as object — needs converting to datetime

#### Explore transactions table 

In [5]:
print("=" * 40)
print("TRANSACTIONS DATASET")
print("=" * 40)
print(transactions_df.info())
print("\nMissing values:")
print(transactions_df.isnull().sum())
print("\nDuplicates:", transactions_df.duplicated().sum())
print("\nSample data:")
transactions_df.head()

TRANSACTIONS DATASET
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   transaction_id  10000 non-null  object 
 1   customer_id     10000 non-null  object 
 2   product_id      10000 non-null  object 
 3   quantity        10000 non-null  int64  
 4   unit_price      10000 non-null  float64
 5   total_amount    10000 non-null  float64
 6   date            10000 non-null  object 
 7   store_type      10000 non-null  object 
 8   payment_method  10000 non-null  object 
dtypes: float64(2), int64(1), object(6)
memory usage: 703.3+ KB
None

Missing values:
transaction_id    0
customer_id       0
product_id        0
quantity          0
unit_price        0
total_amount      0
date              0
store_type        0
payment_method    0
dtype: int64

Duplicates: 0

Sample data:


,transaction_id,customer_id,product_id,quantity,unit_price,total_amount,date,store_type,payment_method
0,T00001,C0297,P028,5,325.78,1628.90,2025-06-02,In-Store,Cash
1,T00002,C0922,P001,1,321.52,321.52,2025-04-22,Online,Apple Pay
2,T00003,C0941,P005,1,213.85,213.85,2025-10-26,In-Store,Debit Card
3,T00004,C0719,P004,2,446.63,893.26,2025-08-25,Online,Debit Card
4,T00005,C0729,P027,3,137.15,411.45,2025-04-07,Online,Apple Pay


##### Transaction Insight
- 10,000 transactions with no missing values or duplicates
- quantity and financial columns (unit_price, total_amount) are correct data types
- date stored as object — needs converting to datetime
- Both Online and In-Store channels represented
- 5 payment methods captured

In [6]:
print("Unit Price Range:")
print(f"Minimum price: £{products_df['unit_price'].min()}")
print(f"Maximum price: £{products_df['unit_price'].max()}")
print(f"Average price: £{products_df['unit_price'].mean().round(2)}")


Unit Price Range:
Minimum price: £18.14
Maximum price: £478.82
Average price: £247.8


In [7]:
print(f"Min unit price: £{transactions_df['unit_price'].min()}")
print(f"Max unit price: £{transactions_df['unit_price'].max()}")
print(f"Total revenue: £{transactions_df['total_amount'].sum().round(2)}")
print(f"Average transaction value: £{transactions_df['total_amount'].mean().round(2)}")

Min unit price: £18.14
Max unit price: £478.82
Total revenue: £7425198.71
Average transaction value: £742.52


## Data Cleaning 
Fix date columns in customers and transactions tables

In [8]:
customers_df['join_date'] = pd.to_datetime(customers_df['join_date'])
transactions_df['date'] = pd.to_datetime(transactions_df['date'])


##### Extract useful date features from transactions

In [9]:
transactions_df['year'] = transactions_df['date'].dt.year
transactions_df['month'] = transactions_df['date'].dt.month
transactions_df['month_name'] = transactions_df['date'].dt.strftime('%B')
transactions_df['day_of_week'] = transactions_df['date'].dt.strftime('%A')

##### Verify

In [10]:
print("Customers join_date dtype:", customers_df['join_date'].dtype)
print("Transactions date dtype:", transactions_df['date'].dtype)
print("\n✅ Date columns fixed successfully!")
transactions_df.head()

Customers join_date dtype: datetime64[ns]
Transactions date dtype: datetime64[ns]

✅ Date columns fixed successfully!


,transaction_id,customer_id,product_id,quantity,unit_price,total_amount,date,store_type,payment_method,year,month,month_name,day_of_week
0,T00001,C0297,P028,5,325.78,1628.90,2025-06-02,In-Store,Cash,2025,6,June,Monday
1,T00002,C0922,P001,1,321.52,321.52,2025-04-22,Online,Apple Pay,2025,4,April,Tuesday
2,T00003,C0941,P005,1,213.85,213.85,2025-10-26,In-Store,Debit Card,2025,10,October,Sunday
3,T00004,C0719,P004,2,446.63,893.26,2025-08-25,Online,Debit Card,2025,8,August,Monday
4,T00005,C0729,P027,3,137.15,411.45,2025-04-07,Online,Apple Pay,2025,4,April,Monday


##### Adding profit column

In [11]:
transactions_df = transactions_df.merge(
    products_df[['product_id', 'cost_price']], 
    on='product_id', how='left')

transactions_df['profit'] = round(
    transactions_df['total_amount'] - 
    (transactions_df['cost_price'] * transactions_df['quantity']), 2)

print("✅ Profit column added!")
transactions_df.head()

✅ Profit column added!


,transaction_id,customer_id,product_id,quantity,unit_price,total_amount,date,store_type,payment_method,year,month,month_name,day_of_week,cost_price,profit
0,T00001,C0297,P028,5,325.78,1628.90,2025-06-02,In-Store,Cash,2025,6,June,Monday,189.84,679.70
1,T00002,C0922,P001,1,321.52,321.52,2025-04-22,Online,Apple Pay,2025,4,April,Tuesday,131.02,190.50
2,T00003,C0941,P005,1,213.85,213.85,2025-10-26,In-Store,Debit Card,2025,10,October,Sunday,87.45,126.40
3,T00004,C0719,P004,2,446.63,893.26,2025-08-25,Online,Debit Card,2025,8,August,Monday,190.30,512.66
4,T00005,C0729,P027,3,137.15,411.45,2025-04-07,Online,Apple Pay,2025,4,April,Monday,93.40,131.25


In [12]:
transactions_df['profit'] = round(
    transactions_df['total_amount'] - 
    (transactions_df['cost_price'] * transactions_df['quantity']), 2)

transactions_df['profit_margin'] = round(
    (transactions_df['profit'] / transactions_df['total_amount']) * 100, 2)


In [13]:
print(" Profit and Profit Margin columns added!")
print(f"\nAverage profit margin: {transactions_df['profit_margin'].mean().round(2)}%")
transactions_df[['transaction_id', 'total_amount', 'cost_price', 'profit', 'profit_margin']].head()

 Profit and Profit Margin columns added!

Average profit margin: 47.75%


,transaction_id,total_amount,cost_price,profit,profit_margin
0,T00001,1628.90,189.84,679.70,41.73
1,T00002,321.52,131.02,190.50,59.25
2,T00003,213.85,87.45,126.40,59.11
3,T00004,893.26,190.30,512.66,57.39
4,T00005,411.45,93.40,131.25,31.90


#### Profit Margin Analysis 

In [14]:
print("=" * 40)
print("PROFIT MARGIN ANALYSIS")
print("=" * 40)
print(f"Average profit margin: {transactions_df['profit_margin'].mean().round(2)}%")
print(f"Highest profit margin: {transactions_df['profit_margin'].max().round(2)}%")
print(f"Lowest profit margin: {transactions_df['profit_margin'].min().round(2)}%")
print(f"\nTotal Revenue: £{transactions_df['total_amount'].sum().round(2):,}")
print(f"Total Cost: £{(transactions_df['cost_price'] * transactions_df['quantity']).sum().round(2):,}")
print(f"Total Profit: £{transactions_df['profit'].sum().round(2):,}")

PROFIT MARGIN ANALYSIS
Average profit margin: 47.75%
Highest profit margin: 59.8%
Lowest profit margin: 30.81%

Total Revenue: £7,425,198.71
Total Cost: £3,924,240.72
Total Profit: £3,500,957.99


#### Profit margin insight
- Quantra Retail maintains a healthy average profit margin of **47.75%**
- This means for every £1 of revenue generated, approximately 
**48 pence is profit**
- Total revenue of **£7.4M** generated **£3.5M** in profit 
across 10,000 transactions
- Margin varies across transactions due to differences in product 
categories and cost structures
- Further analysis in SQL and Power BI will reveal which product 
categories and regions are most profitable

##### Save Cleaned Datasets
Saving the cleaned and enriched datasets as new CSV files ready for SQL import and Power BI analysis. Changes made during cleaning:
- Converted `join_date` to datetime format
- Converted `date` to datetime format
- Extracted `year`, `month`, `month_name`, `day_of_week` from date
- Added `cost_price` from products table
- Calculated `profit` column
- Calculated `profit_margin` percentage column

In [15]:
products_df.to_csv('../data/products_cleaned.csv', index=False)
customers_df.to_csv('../data/customers_cleaned.csv', index=False)
transactions_df.to_csv('../data/transactions_cleaned.csv', index=False)

print("✅ products_cleaned.csv saved!")
print("✅ customers_cleaned.csv saved!")
print("✅ transactions_cleaned.csv saved!")
print("\n Data cleaning complete! Ready for SQL analysis.")

✅ products_cleaned.csv saved!
✅ customers_cleaned.csv saved!
✅ transactions_cleaned.csv saved!

 Data cleaning complete! Ready for SQL analysis.


In [16]:
!pip install pymysql sqlalchemy


[notice] A new release of pip is available: 26.0.1 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


#### Import cleaned data to MYSQL

In [17]:
from sqlalchemy import create_engine

engine = create_engine('mysql+pymysql://root:password123@localhost/quantra_retail')

products_df.to_sql('products', engine, if_exists='replace', index=False)
customers_df.to_sql('customers', engine, if_exists='replace', index=False)
transactions_df.to_sql('transactions', engine, if_exists='replace', index=False)

print("✅ products table imported!")
print("✅ customers table imported!")
print("✅ transactions table imported!")
print("\n🎉 All tables ready in MySQL!")

✅ products table imported!
✅ customers table imported!
✅ transactions table imported!

🎉 All tables ready in MySQL!
